rfcs/0007-synchronous-commit.md

```diff
@@ -0,0 +1,442 @@
+
+If we successfully write a key that succesfully persisted to S3, at this point, all these watermarks will point to the same position. 
+
+For reads in a transaction, it's possible to allow users specify the watermark as `ReadWatermark::LastRemotePersisted` or `ReadWatermark::LastLocalPersisted`. This might let users read an older version of the data, but it won't violate the semantics of conflict checking: if there exists a newer version of the same key, this transaction could be rolled back on commit.
+
+In conclusion, the difference between the current model and this proposal is adding a "Committed" watermark. And as it does not make sense to use `DurabilityLevel` to contain it, we propose to use `ReadWatermark` to identify the positions of the commit history for read operations.
+
+In most of the use cases, users do not need to care about this `ReadWatermark` option. `ReadWatermark::LastCommitted` is the default and the most commonly used value, all the read are safely committed, and the read operations will never block each other. At some mission critical use cases (like using SlateDB as a Kafka-like at-least-once data bus), users may do not want to touch any unpersisted data which risks data loss, and they can set the watermark as `ReadWatermark::LastLocalPersisted` or `ReadWatermark::LastRemotePersisted` to achieve this goal.
+
+In the earlier draft, I've put a `ReadWatermark::LastCommitting` in the `ReadWatermark` enum. But after several discussions, we gradually reached a consensus that it's not a good idea to keep it, because it doesn't improve the performance of read operations, but only allows reading data that might be rolled back later, which might violate the semantics of conflict checking, and also introduce more corner cases to handle.
+
+### Sync Commit
+
+In contrast to the read side which may mix the positions in term of "durability" and "commit", the write / commit side is more straightforward, it only need to track the things about durability.
+
+To implement Synchronous Commit semantics, the write side remains do not need to change at all: the operation returns once the WAL is persisted according to the specified durability level. A `DurabilityLevel` is all we need here.
+
+Like:
+
+```rust
+let opts = WriteOptions::new().with_await_durability(DurabilityLevel::Remote);
+db.write(key, value, opts).await?;
+```
+
+In other systems, this option is often named as `sync`, to emphasize the idea of synchronous commit. To respect the idea of Least Surprise, it might be not a bad idea to rename `await_durability` to `sync`.
+
+```rust
+let opts = WriteOptions::new().with_sync(DurabilityLevel::Memory);
+db.write(key, value, opts).await?;
+```
+
+However, for the writes that not expected to durablely persisted, this can be thought of as **disabling** synchronous commit entirely rather than synchronizing to memory. The code sample may be like:
+
+```rust
+enum SyncLevel {
+    Off,
```

Comment:
I'm trying to understand the relationship between `SyncLevel::Off` and `LastCommitted`. If a writer sets the sync to off, and reader sets to last committed, is there a window of time where a write might be invisible to the reader since the write might not yet be committed? Or is off on the writer side the same thing as last committed on the reader side?

Reply:
i think a window still exists even it's `sync=Off`, the example in https://github.com/slatedb/slatedb/pull/440#discussion_r1966518506 can also reflects this behaviour.

a write with `sync=Off` does not means it will be committed immediately, there might also have chance to exists previous writes with `sync=Remote` which is still inflight to commit. at this time, the later write with `sync=Off` should await the earlier write with `sync=Remote` to be committed 🤔.

on the other hand, if all the writes are marked with `sync=Off`, these writes will be commited immediately, we can put "append to WAL; apply to MemTable" in an atomic operation for this case here.